### PR TITLE
Cleanup backcompat code kpo

### PR DESCRIPTION
**Pull request by dstandish**
Assumes https://github.com/apache/airflow/pull/27524 will be merged first!

I did a thorough review of the kubernetes system KPO backcompat test.

1. most of the time, no actual backcompat logic was tested and the tests were simply duplicates that can be removed.
2. sometimes actual backcompat was tested and in this case i moved to the main KPO system test module and indicated it is a test of a backcompat feature, and where possible i parametrized.
3. in a some cases, tests that needed to be retained were not actually system tests so i either moved to the unit test module or made it a real system test.
4. in the retained KPO system test module I had to convert it from unittests

If you're wondering "what did you do with all the _backcompat tests" ... you can review the commit history.  For the most part, each commit is a distinct change / fix, e.g. moving or deleting a test, as i worked through the module.

### CODE REVIEW

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 1]
```diff
@@ -62,6 +62,23 @@
     from airflow.utils.context import Context
 
 
+def _task_id_to_pod_name(val):
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: this is from https://github.com/apache/airflow/pull/27524 which should be merged first

File: kubernetes_tests/test_kubernetes_pod_operator.py [hunk 2]
```diff
@@ -79,17 +77,14 @@ def get_kubeconfig_path():
 
 def get_label():
     test = os.environ.get("PYTEST_CURRENT_TEST")
+    test = test.split(" ")[0]  # if invoked from setup, will have suffix ' (setup)', e.g.
```
Comment (dstandish) [inline kubernetes_tests/test_kubernetes_pod_operator.py]: this change was necessitated by the switch to pytest... the test name shows up differently in diff contexts...

File: kubernetes_tests/test_kubernetes_pod_operator.py [hunk 3]
```diff
@@ -79,17 +77,14 @@ def get_kubeconfig_path():
...
     label = "".join(filter(str.isalnum, test)).lower()
-    return label[-63]
+    return label[-63:]
```
Comment (dstandish) [inline kubernetes_tests/test_kubernetes_pod_operator.py:82]: ... as a result, looking at just this one character was unreliable. here we also make it longer so it's more meaningful.

File: kubernetes_tests/test_kubernetes_pod_operator.py [hunk 4]
```diff
@@ -79,17 +77,14 @@ def get_kubeconfig_path():
...
 
 
 @pytest.mark.execution_timeout(180)
-class TestKubernetesPodOperatorSystem(unittest.TestCase):
-    def get_current_task_name(self):
-        # reverse test name to make pod name unique (it has limited length)
-        return "_" + unittest.TestCase.id(self).replace(".", "_")[::-1]
```
Comment (dstandish) [inline kubernetes_tests/test_kubernetes_pod_operator.py:90]: if we're just after "unique" we can use something simpler, like uuid4

File: kubernetes_tests/test_kubernetes_pod_operator.py [hunk 5]
```diff
@@ -51,7 +50,7 @@
 
 def create_context(task) -> Context:
     dag = DAG(dag_id="dag")
-    tzinfo = pendulum.timezone("Europe/Amsterdam")
+    tzinfo = pendulum.timezone("Europe/Amsterdam")  # type: ignore
```
Comment (uranusjr) [inline kubernetes_tests/test_kubernetes_pod_operator.py]: Hmm, why is this needed, is pendulum’s type annotation falty?

Comment (dstandish) [reply kubernetes_tests/test_kubernetes_pod_operator.py]: yea... sigh... so... i was getting `module is not callable` error. happened twice  in a row so not flakey.  maybe it was confused by the other `timezone`. it doesn't make any sense. but it's just a test so just ... went the easy route.  and frankly i have no clue how i would fix it otherwise.

Comment (uranusjr) [reply kubernetes_tests/test_kubernetes_pod_operator.py]: Ah that makes sense, there _is_ a `pendulum.timezone` subpackage.

Some of our code uses `pendulum.tz.timezone` instead and that seems to work fine.

Comment (dstandish) [reply kubernetes_tests/test_kubernetes_pod_operator.py]: that's so weird though... how does that even work, if there is a function and a package that have the same name?  is it becaues it's a package and not a module that it works? and they just say "tie goes to the function"?

Comment (uranusjr) [reply kubernetes_tests/test_kubernetes_pod_operator.py]: It’s because Pendulum imports `timezone` in `__init__.py`, and then later overrides the name to a function:

https://github.com/sdispater/pendulum/blob/cc2db5d573f8936a475148bda3ff92057723fc48/pendulum/__init__.py

It is not unlike how Mypy dislikes this:

```python
a = 1
a = "2"  # type: ignore

def f(x: str) -> None:
    print(x)

f(a)  # error: Argument 1 to "f" has incompatible type "int"; expected "str"
```

Pendulum does not really care about strong typing, so everything’s fine from its perspective, but from Mypy’s perspective `pendulum.timezone` is still declared as a module (because that’s what appears first), not a function.

Comment (dstandish) [reply kubernetes_tests/test_kubernetes_pod_operator.py]: oh, yeah, makes sense.  bad pendulum.

Comment (dstandish) [reply kubernetes_tests/test_kubernetes_pod_operator.py]: ok i updated to use `tz`

File: kubernetes_tests/test_kubernetes_pod_operator.py (hunk 2 above)
Comment (uranusjr) [inline kubernetes_tests/test_kubernetes_pod_operator.py]:
```suggestion
    test = test.split(" ", 1)[0]  # if invoked from setup, will have suffix ' (setup)', e.g.
```

No need to do additional splits.

File: kubernetes_tests/test_kubernetes_pod_operator.py [hunk 6]
```diff
@@ -387,8 +369,53 @@ def test_pod_resources(self):
...
+                                        k8s.V1NodeSelectorRequirement(
+                                            key="beta.kubernetes.io/os",
+                                            operator="In",
+                                            values=["linux"],
+                                        )
+                                    ]
+                                )
+                            ]
+                        )
+                    )
```
Comment (uranusjr) [inline kubernetes_tests/test_kubernetes_pod_operator.py:391]: OMG this is ugly (not your fault)

Comment (dstandish) [reply kubernetes_tests/test_kubernetes_pod_operator.py:391]: yes, quite. makes you wonder whether it really makes sense to deprecate the old way

Comment (potiuk) [reply kubernetes_tests/test_kubernetes_pod_operator.py:391]: Ugly it is. I agree. What are the benefits in deprecating the old way BTW :)

Comment (dstandish) [reply kubernetes_tests/test_kubernetes_pod_operator.py:391]: well, this was before my time, but obvs if you're using objects it can be helpful e.g. with autocomplete and validation.  
but yeah it was actually kind of tricky to use. 

maybe we should only be deprecating the airflow objects, and still support json or k8s object.

but, note that this PR doesn't change anything related to deprecations.  it just gets rid of noise in the "backcompat" test, and moves legitimate "backcompat" tests inline with the main tests. which helps e.g. in this case becasue you can see side-by-side new vs "backcompat". in this particular test, there actually _wasn't_ any test of the "new" behavior previously (maybe because it's such a pain in the ass?) so i added that here.

anyway, if any of you "looky lous" has an inclination to approve that would be welcome 😛

**Review by jedcunningham — APPROVED** [review]

**Review by potiuk — APPROVED** [review]
